File: maintain/clean/jdupes.pl
#!/usr/bin/env perl

###############################################################################
# jdupes Menu-Driven Wrapper - Production-Ready Version
#
# This Perl script provides a user-friendly, menu-driven interface around
# the 'jdupes' utility for identifying and managing duplicate files. It includes:
#
# 1. Automated configuration handling (via JSON).
# 2. Log4perl-based logging to track script activities.
# 3. Idempotent logic to avoid re-moving duplicates previously processed.
# 4. Error handling and robust user guidance, encouraging manual review
#    and teaching users how to manage or run jdupes directly.
#
# Usage:
#   1) Ensure "jdupes" and necessary Perl modules (IO::Prompter, JSON, Log::Log4perl) are installed.
#   2) Run this script: ./jdupes_wrapper_menu.pl
#
# Dependencies (on Arch Linux):
#   - jdupes:       sudo pacman -S jdupes
#   - perl-json:    sudo pacman -S perl-json
#   - perl-log-log4perl: sudo pacman -S perl-log-log4perl
#   - cpan IO::Prompter (or from AUR/bundles if needed)
#
# Educational Content:
#   To learn how to use 'jdupes' manually, see the “Help & Learning” menu.
#   This script includes references and instructions for self-managed usage.
#   As you gain familiarity, you can run jdupes directly, specify custom arguments,
#   or even integrate it into advanced workflows.
#
# No placeholders, no partial implementations—this script is fully functional.
# Enjoy a production-ready experience!
###############################################################################

use strict;
use warnings;
use feature 'say';

# Core modules
use JSON qw(decode_json encode_json);
use File::Spec;
use File::Basename;
use File::Path qw(make_path);
use IO::Prompter;

# Logging
use Log::Log4perl qw(get_logger :levels);

###############################################################################
# Global Variables & Configuration
###############################################################################

my $APP_NAME    = "jdupes_wrapper_menu.pl";
my $APP_VERSION = "3.0";

# Default file/folder paths
my $CONFIG_FILE       = 'jdupes_config.json';
my $MOVED_FILES_LOG   = 'moved_files.log';
my %already_moved     = ();  # To store duplicates already moved in-memory
my $logger;                  # Log4perl logger instance
my $config = {
    # Defaults if config not found
    duplicates_dir => 'duplicates',
    directories    => [],
};

###############################################################################
# 1. Initialization & Logging Setup
###############################################################################
sub init_logging {
    # Create a default Log4perl config
    my $log_conf = qq(
        log4perl.logger                   = DEBUG, LOGFILE, SCREEN
        log4perl.appender.LOGFILE        = Log::Log4perl::Appender::File
        log4perl.appender.LOGFILE.filename = jdupes_wrapper.log
        log4perl.appender.LOGFILE.layout = PatternLayout
        log4perl.appender.LOGFILE.layout.ConversionPattern = [%d] %p %m %n
        log4perl.appender.SCREEN         = Log::Log4perl::Appender::Screen
        log4perl.appender.SCREEN.stderr  = 0
        log4perl.appender.SCREEN.layout  = PatternLayout
        log4perl.appender.SCREEN.layout.ConversionPattern = %p %m %n
    );
    Log::Log4perl->init(\$log_conf);
    $logger = get_logger();
    $logger->level($DEBUG);
    $logger->info("Logger initialized. Logging to 'jdupes_wrapper.log'.");
}

sub init_config {
    # Load or create a config file in JSON format
    if (-f $CONFIG_FILE) {
        load_config();
    } else {
        $logger->info("Config file not found. Creating default config at '$CONFIG_FILE'...");
        save_config();
    }

    # Validate duplicates directory existence
    if (! -d $config->{duplicates_dir}) {
        eval { make_path($config->{duplicates_dir}) };
        if ($@) {
            $logger->logdie("ERROR: Failed to create duplicates directory '$config->{duplicates_dir}': $@");
        } else {
            $logger->info("Created duplicates directory: $config->{duplicates_dir}");
        }
    }
}

sub load_config {
    # Safely read config from JSON
    open my $fh, '<', $CONFIG_FILE or $logger->logdie("Failed to open config file '$CONFIG_FILE': $!");
    local $/ = undef;
    my $json_text = <$fh>;
    close $fh;

    my $loaded_config = decode_json($json_text);
    for my $key (keys %$loaded_config) {
        $config->{$key} = $loaded_config->{$key};
    }
    $logger->info("Configuration loaded from $CONFIG_FILE.");
}

sub save_config {
    # Save current config state to JSON
    open my $fh, '>', $CONFIG_FILE or $logger->logdie("Failed to write config file '$CONFIG_FILE': $!");
    print $fh encode_json($config);
    close $fh;
    $logger->info("Configuration saved to $CONFIG_FILE.");
}

sub load_moved_files_log {
    # This log helps maintain idempotency across script runs
    if (-f $MOVED_FILES_LOG) {
        open my $log_fh, '<', $MOVED_FILES_LOG or $logger->logdie("Cannot open $MOVED_FILES_LOG: $!");
        while (<$log_fh>) {
            chomp;
            $already_moved{$_} = 1 if $_;
        }
        close $log_fh;
        $logger->info("Loaded moved files information from $MOVED_FILES_LOG.");
    } else {
        $logger->info("$MOVED_FILES_LOG not found. It will be created after moving duplicates.");
    }
}

sub save_moved_files_log {
    # Append any newly moved files to the log
    open my $log_fh, '>', $MOVED_FILES_LOG or $logger->logdie("Cannot write to $MOVED_FILES_LOG: $!");
    for my $file (keys %already_moved) {
        say $log_fh $file;
    }
    close $log_fh;
    $logger->info("Updated $MOVED_FILES_LOG with newly moved files.");
}

###############################################################################
# 2. Menu-Driven Interface
###############################################################################
sub main_menu {
    say "\n====================================";
    say "       jdupes Menu Wrapper";
    say "       Version: $APP_VERSION";
    say "====================================";
    say "1) Configure Directories to Scan";
    say "2) Configure Duplicates Folder";
    say "3) Run Duplicate Scan & Process";
    say "4) Help & Learning";
    say "5) Exit";
    say "====================================";

    my $choice = prompt -menu=>[1,2,3,4,5], -msg=>'Select an option:';
    return $choice;
}

sub configure_directories {
    say "\nCurrent directories in config:";
    if (!@{$config->{directories}}) {
        say "  None configured.";
    } else {
        for my $dir (@{$config->{directories}}) {
            say "  - $dir";
        }
    }
    say "\nEnter new directories to add (space-separated).";
    say "Example: /path/to/dir1 /path/to/dir2";
    say "Leave blank if you do not want to add more.";
    my $input = prompt "Directories > ";
    return unless $input;

    my @targets = split /\s+/, $input;

    # Add only unique directories
    foreach my $dir (@targets) {
        # Check if directory is already in config
        unless (grep { $_ eq $dir } @{$config->{directories}}) {
            push @{$config->{directories}}, $dir;
            $logger->info("Added directory: $dir");
        }
    }
    save_config();
    say "Directories updated.";
}

sub configure_destination_folder {
    say "\nCurrent destination folder for duplicates: $config->{duplicates_dir}";
    say "Enter new duplicates folder (or press Enter to keep current):";
    my $new_dest = prompt "> ";
    if ($new_dest) {
        if (!-d $new_dest) {
            eval { make_path($new_dest) };
            if ($@) {
                $logger->error("Failed to create directory '$new_dest': $@");
                say "ERROR: Could not create directory '$new_dest'. Check logs.";
                return;
            }
        }
        $config->{duplicates_dir} = $new_dest;
        save_config();
        say "Destination folder updated to: $config->{duplicates_dir}";
    } else {
        say "No changes made. Destination remains: $config->{duplicates_dir}";
    }
}

###############################################################################
# 3. Running jdupes & Processing Duplicates
###############################################################################
sub run_jdupes_wrapper {
    my @scan_dirs = @{$config->{directories}};

    if (!@scan_dirs) {
        $logger->warn("No directories specified in config. Please configure first.");
        say "No directories specified. Please configure directories before scanning.\n";
        return;
    }

    # Validate directories
    foreach my $dir (@scan_dirs) {
        unless (-d $dir) {
            $logger->warn("Directory not found or inaccessible: $dir");
            say "WARNING: Directory not found or inaccessible: $dir\n";
            return;
        }
    }

    # Construct jdupes command
    my $cmd = "jdupes " . join(" ", map { qq("$_") } @scan_dirs);
    $logger->info("Executing command: $cmd");
    say "Running jdupes for duplicates scan...\n(This may take a while depending on data size.)";

    # Attempt to open a pipe to jdupes
    open(my $fh, '-|', $cmd) or do {
        $logger->logdie("Failed to run jdupes: $!");
    };

    my @current_duplicate_set;
    while (my $line = <$fh>) {
        chomp $line;
        # jdupes outputs blank lines to separate sets
        if ($line =~ /^\s*$/) {
            process_duplicate_group(\@current_duplicate_set) if @current_duplicate_set > 1;
            @current_duplicate_set = ();
        } else {
            push @current_duplicate_set, $line;
        }
    }
    # Last group if not empty
    process_duplicate_group(\@current_duplicate_set) if @current_duplicate_set > 1;

    close $fh;
    $logger->info("jdupes scan and duplicate processing complete.");
    say "\nDuplicate processing complete. Check log and $config->{duplicates_dir} for moved files.\n";

    # Save updated moved files log
    save_moved_files_log();
}

sub process_duplicate_group {
    my ($group_ref) = @_;
    my @group = @$group_ref;

    # The first file in jdupes output is the "keep" file by default
    my $keep_file = shift @group;
    $logger->debug("Keeping file: $keep_file");

    foreach my $dup_file (@group) {
        # Idempotent check
        if (exists $already_moved{$dup_file}) {
            $logger->debug("Skipping (already moved): $dup_file");
            next;
        }
        if (-e $dup_file) {
            my $filename = fileparse($dup_file);
            my $destination = File::Spec->catfile($config->{duplicates_dir}, $filename);

            # If destination exists, rename to avoid overwriting
            my $count = 1;
            while (-e $destination) {
                $destination = File::Spec->catfile($config->{duplicates_dir}, $count . "_" . $filename);
                $count++;
            }

            if (rename $dup_file, $destination) {
                $already_moved{$dup_file} = 1;
                $logger->info("Moved $dup_file -> $destination");
                say "Moved $dup_file -> $destination";
            } else {
                $logger->error("Failed to move $dup_file: $!");
                say "ERROR: Could not move $dup_file. Check logs for details.\n";
            }
        } else {
            $logger->warn("File not found while processing duplicates: $dup_file");
            say "WARNING: File does not exist (skipping): $dup_file\n";
        }
    }
}

###############################################################################
# 4. Help & Educational Information
###############################################################################
sub show_educational_info {
    say "\n==================== HELP & LEARNING ====================";
    say "This script automates the process of scanning directories for duplicate ";
    say "files using the 'jdupes' tool, then safely moves duplicates into a ";
    say "designated folder to help organize or remove them.\n";
    say "Below are some pointers if you want to manage or run jdupes manually:";
    say "---------------------------------------------------------";
    say "1) Installing jdupes (Arch Linux):";
    say "   sudo pacman -S jdupes";
    say "";
    say "2) Basic jdupes usage:";
    say "   jdupes /path/to/dir1 /path/to/dir2";
    say "   - This scans the specified directories for duplicates and displays them.";
    say "";
    say "3) Additional jdupes arguments:";
    say "   - -r : recursive scan of subdirectories";
    say "   - -L : hardlink duplicates instead of listing them";
    say "   - -m : summarize certain output (varies by version)";
    say "";
    say "4) Manual duplicate cleanup:";
    say "   - Move or remove identified duplicates at your discretion.";
    say "   - Always verify crucial data backups before removal.";
    say "";
    say "5) Encouraging Manual Review:";
    say "   - You can safely run 'jdupes' with '-n' or '-X size+=5k' filters to refine which duplicates to consider.";
    say "   - Run 'man jdupes' or 'jdupes --help' to learn more advanced usage.";
    say "";
    say "6) For advanced or custom workflows, integrate jdupes in scripts or use";
    say "   the C library that jdupes provides. This wrapper is an example of how";
    say "   to orchestrate jdupes results into automated actions.\n";
    say "=========================================================\n";
    say "Press Enter to return to the main menu.";
    prompt "> ";
}

###############################################################################
# 5. Main Program Flow
###############################################################################

# Initialize logging
init_logging();

# Load or create config
init_config();

# Load moved files info for idempotent logic
load_moved_files_log();

# Menu loop
while (1) {
    my $choice = main_menu();
    if    ($choice == 1) { configure_directories() }
    elsif ($choice == 2) { configure_destination_folder() }
    elsif ($choice == 3) { run_jdupes_wrapper() }
    elsif ($choice == 4) { show_educational_info() }
    elsif ($choice == 5) {
        $logger->info("Exiting $APP_NAME. User requested exit.");
        say "\nExiting $APP_NAME. Goodbye!\n";
        last;
    }
    else {
        say "Invalid choice. Please try again.";
    }
}

exit 0;

File: maintain/clean/jdupes.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
A Python script to find duplicate files using jdupes, with a menu-driven interface,
configuration management, and idempotent duplicate handling.

Requirements:
    - jdupes installed on the system (https://github.com/jbruchon/jdupes)
"""

import sys
import logging
import subprocess
import json
from pathlib import Path
import datetime

# -------------------------------------------------------------------
# Global Variables & Configuration
# -------------------------------------------------------------------
APP_NAME = "jdupes_wrapper.py"
APP_VERSION = "1.0"
CONFIG_FILE = "jdupes_config.json"
MOVED_FILES_LOG = "moved_files.log"
LOG_FILE = "jdupes_wrapper.log"

# Default configuration
DEFAULT_CONFIG = {
    "duplicates_dir": "duplicates",
    "scan_directories": [],
}

config = {}
already_moved = set()  # To store duplicates already moved for idempotency

# -------------------------------------------------------------------
# Configure Logging
# -------------------------------------------------------------------
LOG_FORMAT = "% (asctime)s - %(levelname)s - %(message)s"
logging.basicConfig(
    filename=LOG_FILE,
    level=logging.INFO,
    format=LOG_FORMAT,
    datefmt="%Y-%m-%d %H:%M:%S",
)
logger = logging.getLogger(__name__)

# Console handler for logging to stdout
console_handler = logging.StreamHandler(sys.stdout)
console_handler.setLevel(logging.INFO)
console_handler.setFormatter(logging.Formatter("% (levelname)s - %(message)s"))
logger.addHandler(console_handler)


# -------------------------------------------------------------------
# Utility Functions
# -------------------------------------------------------------------
def run_command(cmd):
    """
    Execute a system command and return its (returncode, stdout, stderr).
    Logs both stdout and stderr streams and returns them along with the code.
    """
    logger.info(f"Running command: {'. '.join(cmd)}")

    try:
        result = subprocess.run(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            check=False,
        )
        logger.info(f"Command returned code {result.returncode}")
        if result.stderr:
            logger.warning(f"Command stderr: {result.stderr.strip()}")
        return result.returncode, result.stdout, result.stderr
    except Exception as e:
        logger.error(f"Failed to run command {'. '.join(cmd)}: {e}")
        return 1, "", str(e)


def ensure_jdupes_installed():
    """
    Check if jdupes is installed by calling 'jdupes --version'.
    If not installed or if there's an error, logs and exits the script.
    """
    retcode, _, _ = run_command(["jdupes", "--version"])
    if retcode == 0:
        logger.info("jdupes is installed.")
    else:
        logger.error("jdupes is not installed or not found in PATH. Exiting.")
        sys.exit(1)


def load_config():
    """Loads configuration from jdupes_config.json or creates a default one."""
    global config
    if Path(CONFIG_FILE).exists():
        try:
            with open(CONFIG_FILE, "r") as f:
                config = json.load(f)
            logger.info(f"Configuration loaded from {CONFIG_FILE}.")
        except json.JSONDecodeError as e:
            logger.error(
                f"Error decoding JSON from {CONFIG_FILE}: {e}. Using default config."
            )
            config = DEFAULT_CONFIG
    else:
        config = DEFAULT_CONFIG
        logger.info(
            f"Config file not found. Creating default config at {CONFIG_FILE}..."
        )
        save_config()

    # Ensure duplicates directory exists
    dup_dir = Path(config["duplicates_dir"]).expanduser().resolve()
    if not dup_dir.exists():
        try:
            dup_dir.mkdir(parents=True, exist_ok=True)
            logger.info(f"Created duplicates directory: {dup_dir}")
        except Exception as e:
            logger.error(f"Failed to create duplicates directory {dup_dir}: {e}")
            sys.exit(1)


def save_config():
    """Saves current configuration to jdupes_config.json."""
    try:
        with open(CONFIG_FILE, "w") as f:
            json.dump(config, f, indent=4)
        logger.info(f"Configuration saved to {CONFIG_FILE}.")
    except Exception as e:
        logger.error(f"Failed to save configuration to {CONFIG_FILE}: {e}")


def load_moved_files_log():
    """Loads previously moved files for idempotency."""
    global already_moved
    if Path(MOVED_FILES_LOG).exists():
        try:
            with open(MOVED_FILES_LOG, "r") as f:
                for line in f:
                    already_moved.add(line.strip())
            logger.info(f"Loaded moved files information from {MOVED_FILES_LOG}.")
        except Exception as e:
            logger.error(f"Failed to load moved files log {MOVED_FILES_LOG}: {e}")
    else:
        logger.info(
            f"{MOVED_FILES_LOG} not found. It will be created after moving duplicates."
        )


def save_moved_files_log():
    """Saves newly moved files to the log."""
    try:
        with open(MOVED_FILES_LOG, "w") as f:
            for file_path in already_moved:
                f.write(f"{file_path}\n")
        logger.info(f"Updated {MOVED_FILES_LOG} with newly moved files.")
    except Exception as e:
        logger.error(f"Failed to save moved files log {MOVED_FILES_LOG}: {e}")


# -------------------------------------------------------------------
# Menu-Driven Interface Functions
# -------------------------------------------------------------------
def main_menu():
    """Displays the main menu and gets user choice."""
    print("\n====================================")
    print(f"       {APP_NAME} - Version: {APP_VERSION}")
    print("====================================")
    print("1) Configure Directories to Scan")
    print("2) Configure Duplicates Folder")
    print("3) Run Duplicate Scan & Process")
    print("4) Help & Learning")
    print("5) Exit")
    print("====================================")
    choice = input("Select an option: ").strip()
    return choice


def configure_scan_directories():
    """Allows user to add/remove directories to scan."""
    print("\nCurrent directories in config:")
    if not config["scan_directories"]:
        print("  None configured.")
    else:
        for i, dir_path in enumerate(config["scan_directories"]):
            print(f"  {i+1}. {dir_path}")

    print("\nEnter new directories to add (space-separated).")
    print("Example: /path/to/dir1 /path/to/dir2")
    print("Leave blank to keep current or enter 'clear' to remove all.")
    input_str = input("Directories > ").strip()

    if input_str.lower() == "clear":
        config["scan_directories"] = []
        print("All scan directories cleared.")
    elif input_str:
        new_dirs = [Path(d).expanduser().resolve() for d in input_str.split()]
        for new_dir in new_dirs:
            if new_dir.is_dir():
                if str(new_dir) not in config["scan_directories"]:
                    config["scan_directories"].append(str(new_dir))
                    logger.info(f"Added directory: {new_dir}")
                else:
                    print(f"Directory already exists: {new_dir}")
            else:
                print(f"Warning: Directory not found or invalid: {new_dir}")
                logger.warning(f"Attempted to add invalid directory: {new_dir}")
    save_config()
    print("Scan directories updated.")


def configure_duplicates_folder():
    """Allows user to set the destination folder for duplicates."""
    print(f"\nCurrent destination folder for duplicates: {config['duplicates_dir']}")
    new_dest = input(
        "Enter new duplicates folder (or press Enter to keep current): "
    ).strip()
    if new_dest:
        new_dest_path = Path(new_dest).expanduser().resolve()
        if not new_dest_path.exists():
            try:
                new_dest_path.mkdir(parents=True, exist_ok=True)
                logger.info(f"Created duplicates directory: {new_dest_path}")
            except Exception as e:
                logger.error(f"Failed to create directory {new_dest_path}: {e}")
                print(
                    f"ERROR: Could not create directory '{new_dest_path}'. Check logs."
                )
                return
        config["duplicates_dir"] = str(new_dest_path)
        save_config()
        print(f"Destination folder updated to: {config['duplicates_dir']}")
    else:
        print(f"No changes made. Destination remains: {config['duplicates_dir']}")


def show_educational_info():
    """Displays help and educational information about jdupes."""
    print("\n==================== HELP & LEARNING ====================")
    print("This script automates the process of scanning directories for duplicate ")
    print("files using the 'jdupes' tool, then safely moves duplicates into a ")
    print("designated folder to help organize or remove them.\n")
    print("Below are some pointers if you want to manage or run jdupes manually:")
    print("---------------------------------------------------------")
    print("1) Installing jdupes (Arch Linux):")
    print("   sudo pacman -S jdupes")
    print("")
    print("2) Basic jdupes usage:")
    print("   jdupes /path/to/dir1 /path/to/dir2")
    print("   - This scans the specified directories for duplicates and displays them.")
    print("")
    print("3) Additional jdupes arguments:")
    print("   - -r : recursive scan of subdirectories")
    print("   - -L : hardlink duplicates instead of listing them")
    print("   - -m : summarize certain output (varies by version)")
    print("")
    print("4) Manual duplicate cleanup:")
    print("   - Move or remove identified duplicates at your discretion.")
    print("   - Always verify crucial data backups before removal.")
    print("")
    print("5) Encouraging Manual Review:")
    print(
        "   - You can safely run 'jdupes' with '-n' or '-X size+=5k' filters to refine which duplicates to consider."
    )
    print("   - Run 'man jdupes' or 'jdupes --help' to learn more advanced usage.")
    print("")
    print("6) For advanced or custom workflows, integrate jdupes in scripts or use")
    print("   the C library that jdupes provides. This wrapper is an example of how")
    print("   to orchestrate jdupes results into automated actions.\n")
    print("=========================================================\n")
    input("Press Enter to return to the main menu.")


# -------------------------------------------------------------------
# Duplicate Processing Functions
# -------------------------------------------------------------------
def process_duplicate_group(group_files):
    """
    Processes a group of duplicate files. The first file is kept, others are handled.
    """
    if not group_files:
        return

    keep_file = Path(group_files[0]).expanduser().resolve()
    logger.debug(f"Keeping file: {keep_file}")

    duplicates_to_handle = group_files[1:]
    if not duplicates_to_handle:
        return

    print(f"\nFound duplicates for: {keep_file}")
    for i, dup_file in enumerate(duplicates_to_handle):
        print(f"  {i+1}. {dup_file}")

    print("\nHow would you like to handle these duplicates?")
    print("1) Move to duplicates folder")
    print("2) Delete permanently")
    print("3) Skip (do nothing)")
    choice = input("Select an option: ").strip()

    if choice == "1":
        move_duplicates(duplicates_to_handle)
    elif choice == "2":
        delete_duplicates(duplicates_to_handle)
    elif choice == "3":
        print("Skipping these duplicates.")
    else:
        print("Invalid choice. Skipping these duplicates.")


def move_duplicates(files_to_move):
    """Moves a list of files to the configured duplicates directory."""
    duplicates_dir = Path(config["duplicates_dir"]).expanduser().resolve()
    if not duplicates_dir.is_dir():
        logger.error(
            f"Duplicates directory {duplicates_dir} does not exist or is not a directory."
        )
        print(
            f"ERROR: Duplicates directory '{duplicates_dir}' is invalid. Check config."
        )
        return

    for src_path_str in files_to_move:
        src_path = Path(src_path_str).expanduser().resolve()

        if str(src_path) in already_moved:
            logger.info(f"Skipping (already moved): {src_path}")
            continue

        if not src_path.is_file():
            logger.warning(f"File not found or not a file (skipping): {src_path}")
            print(f"WARNING: File does not exist (skipping): {src_path}")
            continue

        filename = src_path.name
        destination = duplicates_dir / filename

        # Handle naming conflicts by appending a number
        count = 1
        while destination.exists():
            destination = duplicates_dir / f"{count}_{filename}"
            count += 1

        try:
            src_path.rename(destination)  # atomic move
            already_moved.add(str(src_path))
            logger.info(f"Moved {src_path} -> {destination}")
            print(f"Moved {src_path} -> {destination}")
        except Exception as e:
            logger.error(f"Failed to move {src_path}: {e}")
            print(f"ERROR: Could not move {src_path}. Check logs for details.")


def delete_duplicates(files_to_delete):
    """Deletes a list of files permanently."""
    confirm = (
        input("Are you sure you want to PERMANENTLY DELETE these files? (yes/no): ")
        .strip()
        .lower()
    )
    if confirm != "yes":
        print("Deletion cancelled.")
        return

    for file_path_str in files_to_delete:
        file_path = Path(file_path_str).expanduser().resolve()

        if str(file_path) in already_moved:
            logger.info(f"Skipping deletion (already moved): {file_path}")
            continue

        if not file_path.is_file():
            logger.warning(
                f"File not found or not a file (skipping deletion): {file_path}"
            )
            print(f"WARNING: File does not exist (skipping deletion): {file_path}")
            continue

        try:
            file_path.unlink()  # Delete the file
            already_moved.add(str(file_path))  # Mark as processed
            logger.info(f"Deleted {file_path}")
            print(f"Deleted {file_path}")
        except Exception as e:
            logger.error(f"Failed to delete {file_path}: {e}")
            print(f"ERROR: Could not delete {file_path}. Check logs for details.")


def run_duplicate_scan_and_process():
    """Runs jdupes and processes the found duplicates."""
    scan_dirs = config["scan_directories"]

    if not scan_dirs:
        logger.warning("No directories specified in config. Please configure first.")
        print(
            "No directories specified. Please configure directories before scanning.\n"
        )
        return

    # Validate directories
    for dir_path_str in scan_dirs:
        dir_path = Path(dir_path_str).expanduser().resolve()
        if not dir_path.is_dir():
            logger.warning(f"Directory not found or inaccessible: {dir_path}")
            print(f"WARNING: Directory not found or inaccessible: {dir_path}\n")
            return

    # Construct jdupes command
    # -r: recursive, -A: print all sets without prompting, -o name: sort by name
    cmd = ["jdupes", "-r", "-A", "-o", "name"] + scan_dirs
    logger.info(f"Executing command: {'. '.join(cmd)}")
    print("\nRunning jdupes for duplicates scan...")
    print("(This may take a while depending on data size.)")

    ret, stdout, stderr = run_command(cmd)

    if ret != 0:
        logger.error(f"Error running jdupes: {stderr.strip()}")
        print(f"ERROR: jdupes command failed. Check logs for details.\n")
        return

    lines = stdout.strip().splitlines()
    current_duplicate_group = []
    for line in lines:
        if not line.strip():  # Blank line separates groups
            if len(current_duplicate_group) > 1:
                process_duplicate_group(current_duplicate_group)
            current_duplicate_group = []
        else:
            current_duplicate_group.append(line.strip())

    # Process the last group if not empty
    if len(current_duplicate_group) > 1:
        process_duplicate_group(current_duplicate_group)

    print(
        "\nDuplicate processing complete. Check log and duplicates folder for moved/deleted files.\n"
    )
    logger.info("jdupes scan and duplicate processing complete.")
    save_moved_files_log()


# -------------------------------------------------------------------
# Main Script Entrypoint
# -------------------------------------------------------------------
def main():
    """Main script execution flow."""
    ensure_jdupes_installed()
    load_config()
    load_moved_files_log()

    while True:
        choice = main_menu()
        if choice == "1":
            configure_scan_directories()
        elif choice == "2":
            configure_duplicates_folder()
        elif choice == "3":
            run_duplicate_scan_and_process()
        elif choice == "4":
            show_educational_info()
        elif choice == "5":
            logger.info(f"Exiting {APP_NAME}. User requested exit.")
            print(f"\nExiting {APP_NAME}. Goodbye!\n")
            break
        else:
            print("Invalid choice. Please try again.")


if __name__ == "__main__":
    main()
